core/dbt/task/build.py

```diff
@@ -4,40 +4,96 @@
 from .test import TestRunner as test_runner
 
 from dbt.contracts.results import NodeStatus
-from dbt.graph import ResourceTypeSelector
-from dbt.exceptions import InternalException
+from dbt.exceptions import RuntimeException, InternalException
+from dbt.graph import (
+    parse_difference,
+    ResourceTypeSelector,
+    SelectionSpec,
+)
 from dbt.node_types import NodeType
+from dbt.task.test import TestSelector
 
 
 class BuildTask(RunTask):
-    """The Build task processes all assets of a given process and
-    attempts to 'build' them in an opinionated fashion. Every resource
-    type outlined in RUNNER_MAP will be processed by the mapped runner class.
+    """The Build task processes all assets of a given process and attempts to
+    'build' them in an opinionated fashion.  Every resource type outlined in
+    RUNNER_MAP will be processed by the mapped runner class.
 
     I.E. a resource of type Model is handled by the ModelRunner which is
-    imported as run_model_runner.
-    """
+    imported as run_model_runner. """
+
+    MARK_DEPENDENT_ERRORS_STATUSES = [NodeStatus.Error, NodeStatus.Fail]
 
     RUNNER_MAP = {
         NodeType.Model: run_model_runner,
         NodeType.Snapshot: snapshot_model_runner,
         NodeType.Seed: seed_runner,
         NodeType.Test: test_runner,
     }
+    ALL_RESOURCE_VALUES = frozenset({x for x in RUNNER_MAP.keys()})
 
-    MARK_DEPENDENT_ERRORS_STATUSES = [NodeStatus.Error, NodeStatus.Fail]
+    def __init__(self, args, config):
+        super().__init__(args, config)
+        if self.args.models:
+            if self.args.select:
+                raise RuntimeException(
+                    '"models" and "select" are mutually exclusive arguments'
+                )
+            if self.args.resource_types:
+                raise RuntimeException(
+                    '"models" and "resource_type" are mutually exclusive '
+                    'arguments'
+                )
+
+    @property
+    def resource_types(self):
+        if self.args.models:
+            return [NodeType.Model]
+
+        if not self.args.resource_types:
+            return list(self.ALL_RESOURCE_VALUES)
+
+        values = set(self.args.resource_types)
```

Comment:
Agree that parity with the `list` task is a good guideline to follow. I have half a mind to just entirely remove support for the `--models` flag in `dbt build`, since it's a totally new command in v0.21 and we shouldn't have to worry about backwards compatibility. I think either is a fair way to go, and I hesitate to make you rewrite a bunch of code given that I think you've done a sensible thing.

FWIW my plan is to just about entirely remove `--models` from our documentation (https://github.com/dbt-labs/docs.getdbt.com/pull/802), except for changelog notes and a note about backwards compatibility in `list`.

Reply:
let's think about our future selves and not support it with `build` from the start. not a huge lift at all!